File: src/editor/ui/tag_tree.rs
use godot::{
    engine::{control::SizeFlags, tree_item::TreeCellMode, DisplayServer, ITree, Tree, TreeItem},
    prelude::*,
};

use crate::tag_dictionary::{TagDictionary, SPLIT_CHAR};

const TAG_PATH_META: &str = "ggt_tag_dictionary_path";

#[derive(GodotClass)]
#[class(tool, init, base = Tree)]
pub struct TagTree {
    base: Base<Tree>,
    /// Whether the tag tree is editable.
    editable: bool,
    /// Whether the tag tree is selectable.
    selectable: bool,
    /// the selected tags
    selected_tags: PackedStringArray,
    /// The tag dictionary to display.
    tag_dictionary: Option<Gd<TagDictionary>>,
}

#[godot_api]
impl ITree for TagTree {
    fn ready(&mut self) {
        self.to_gd().connect(
            StringName::from("button_clicked"),
            Callable::from_object_method(&self.to_gd(), "_on_button_clicked"),
        );
        self.to_gd().connect(
            StringName::from("item_selected"),
            Callable::from_object_method(&self.to_gd(), "_on_item_selected"),
        );
        self.to_gd().connect(
            StringName::from("item_edited"),
            Callable::from_object_method(&self.to_gd(), "_on_tag_edited"),
        );
        self.to_gd().set_v_scroll_enabled(true);
        self.to_gd().set_v_size_flags(SizeFlags::EXPAND_FILL);
        self.render_tree();
    }
}

#[godot_api]
impl TagTree {
    #[signal]
    pub fn tag_path_edited(old_tag: String, new_tag: String);
    #[signal]
    pub fn tag_path_removed(tag: String);
    #[signal]
    pub fn tags_added(tags: PackedStringArray);
    #[signal]
    pub fn tags_removed(tags: PackedStringArray);

    /// Gets whether the tag tree is editable.
    #[func]
    pub fn get_editable(&self) -> bool {
        self.editable
    }

    #[func]
    pub fn get_selectable(&self) -> bool {
        self.selectable
    }

    #[func]
    pub fn get_selected_tags(&self) -> PackedStringArray {
        self.selected_tags.clone()
    }

    /// Gets the tag dictionary to display.
    #[func]
    pub fn get_tag_dictionary(&self) -> Option<Gd<TagDictionary>> {
        self.tag_dictionary.clone()
    }

    #[func]
    pub fn is_path_selected(&self, _path: GString) -> bool {
        self.selected_tags
            .as_slice()
            .iter()
            .any(|x| x.to_string().starts_with(_path.to_string().as_str()))
    }

    /// Sets whether the tag tree is editable.
    #[func]
    pub fn set_editable(&mut self, editable: bool) {
        self.editable = editable;
        self.render_tree();
    }

    /// Sets whether the tag tree is selectable.
    #[func]
    pub fn set_selectable(&mut self, selectable: bool) {
        self.selectable = selectable;
        self.render_tree();
    }

    #[func]
    pub fn set_selected_tags(&mut self, selected_tags: PackedStringArray) {
        self.selected_tags = selected_tags;
        self.render_tree();
    }

    /// Sets the tag dictionary to display.
    #[func]
    pub fn set_tag_dictionary(&mut self, tag_dictionary: Option<Gd<TagDictionary>>) {
        self.tag_dictionary = tag_dictionary;
        self.render_tree();
    }

    #[func]
    fn _on_button_clicked(
        &mut self,
        tree_item: Gd<TreeItem>,
        _column_id: i32,
        id: i32,
        _mouse_button_index: i32,
    ) {
        let tag_path = tree_item.clone().get_meta(StringName::from(TAG_PATH_META));

        if let Some(mut tag_dictionary) = self.tag_dictionary.clone() {
            if id == 1 {
                tag_dictionary.bind_mut().remove_path(tag_path.to_string());
                self.to_gd()
                    .emit_signal(StringName::from("tag_path_removed"), &[tag_path]);

                self.render_tree();
            } else {
                let child_count = tree_item.clone().get_child_count();
                let new_tag_name = String::from("new_tag_") + &child_count.to_string();
                let new_tag = if tag_path.clone().to_string().is_empty() {
                    new_tag_name
                } else {
                    format!("{}{}{}", tag_path.to_string(), SPLIT_CHAR, new_tag_name)
                };

                tag_dictionary
                    .bind_mut()
                    .add_tag(new_tag.clone().to_godot());

                self.render_tree();
            }
        }
    }

    #[func]
    fn _on_item_selected(&mut self) {
        if let Some(item) = self.to_gd().get_selected() {
            DisplayServer::singleton()
                .clipboard_set(item.get_meta(TAG_PATH_META.into()).to_string().into());

            if self.selectable {
                let mut found_tags = PackedStringArray::new();
                let mut tags_added = PackedStringArray::new();
                let mut tags_removed = PackedStringArray::new();
                let tag_path = item.get_meta(TAG_PATH_META.into()).to_string();
                let is_selected = item.is_checked(0);

                if let Some(tag_dictionary) = self.tag_dictionary.clone() {
                    found_tags = tag_dictionary.bind().get_tags_from_path(tag_path.clone().into());
                }

                for tag in found_tags.as_slice().iter() {
                    if is_selected {
                        if let Some(index) = self.selected_tags.find(tag.clone(), Some(0)) {
                            self.selected_tags.remove(index);
                            tags_removed.push(tag.clone());
                        }
                    } else {
                        if !self.selected_tags.contains(tag.clone()) {
                            self.selected_tags.push(tag.clone());
                            tags_added.push(tag.clone());
                        }
                    }
                }

                if !tags_added.clone().is_empty() {
                    self.to_gd().emit_signal("tags_added".into(), &[tags_added.to_variant()]);
                }

                if !tags_removed.clone().is_empty() {
                    self.to_gd().emit_signal("tags_removed".into(), &[tags_removed.to_variant()]);
                }

                self.to_gd().call_deferred("render_tree".into(), &[]);
            }
        }
    }

    #[func]
    fn _on_tag_edited(&mut self) {
        if let Some(edited) = self.to_gd().get_edited() {
            let meta_path = edited.get_meta(TAG_PATH_META.into()).to_string();
            let edited_chunk = edited.get_text(0).to_string();

            if meta_path == edited_chunk {
                return;
            }

            // takes the meta_path except for the last element
            let parent_path = meta_path
                .rsplit_once(SPLIT_CHAR)
                .map(|(a, _)| a)
                .unwrap_or("");
            let new_path = if parent_path.is_empty() {
                edited_chunk.clone()
            } else {
                format!("{}{}{}", parent_path, SPLIT_CHAR, edited_chunk.clone())
            };

            if let Some(mut tag_dictionary) = self.tag_dictionary.clone() {
                tag_dictionary
                    .bind_mut()
                    .update_path(meta_path.clone(), new_path.clone());
            }

            self.to_gd().emit_signal(
                "tag_path_edited".into(),
                &[meta_path.to_variant(), new_path.to_variant()],
            );
            self.render_tree();
        }
    }

    fn render_dictionary(&self, dictionary: Dictionary, mut parent: Gd<TreeItem>, path: GString) {
        let keys = dictionary.keys_array();

        for key in keys.iter_shared() {
            let keystring = key.to_string();

            if let Some(variant) = dictionary.get(keystring.clone()) {
                match Dictionary::try_from_variant(&variant) {
                    Ok(dict) => {
                        let new_path = if path.is_empty() {
                            keystring.clone()
                        } else {
                            format!("{}{}{}", path, SPLIT_CHAR, keystring.clone())
                        };

                        let mut item = parent.call("create_child".into(), &[]).to::<Gd<TreeItem>>();

                        if self.selectable {
                            self.set_tree_item_checkable(
                                item.clone(),
                                path.clone(),
                                keystring.clone().into(),
                            );
                        }

                        if self.editable {
                            item.set_editable(0, true);
                            self.set_tree_item_editable_icon(item.clone());
                        }

                        item.set_meta(TAG_PATH_META.into(), new_path.to_variant());
                        item.set_text(0, keystring.clone().to_godot());
                        item.set_tooltip_text(0, new_path.clone().into());

                        self.render_dictionary(dict, item, new_path.clone().into());
                    }
                    Err(convert_error) => {
                        godot::engine::utilities::printerr(
                            convert_error.to_string().to_variant(),
                            &[],
                        );
                    }
                }
            }
        }
    }

    #[func]
    fn render_tree(&mut self) {
        self.to_gd().clear();

        let mut root = self
            .to_gd()
            .call("create_item".into(), &[])
            .to::<Gd<TreeItem>>();

        root.set_meta(TAG_PATH_META.into(), "".to_variant());

        if let Some(tag_dictionary) = self.tag_dictionary.clone() {
            root.set_text(0, tag_dictionary.get_path());
        }

        if self.selectable {
            self.set_tree_item_checkable(root.clone(), "".into(), "".into());
        }

        if self.editable {
            self.set_tree_item_editable_icon(root.clone());
        }

        if let Some(tag_dictionary) = self.get_tag_dictionary() {
            self.render_dictionary(tag_dictionary.bind().get_tree(), root, "".into())
        }
    }

    fn set_tree_item_checkable(
        &self,
        mut item: Gd<TreeItem>,
        p_current_path: GString,
        key: GString,
    ) {
        let new_path = if p_current_path.is_empty() {
            key.clone().to_string()
        } else {
            format!("{}{}{}", p_current_path, SPLIT_CHAR, key.clone())
        };

        item.set_cell_mode(0, TreeCellMode::CHECK);

        if new_path.is_empty() {
            item.set_checked(0, false);
        } else {
            item.set_checked(0, self.is_path_selected(new_path.into()));
        }
    }

    fn set_tree_item_editable_icon(&self, mut item: Gd<TreeItem>) {
        if let Some(icon) = self.to_gd().get_theme_icon("Add".into()) {
            item.add_button(0, icon);
            item.set_button_tooltip_text(0, 0, "Add tag".into());
        }

        if let Some(icon) = self.to_gd().get_theme_icon("Remove".into()) {
            item.add_button(0, icon);
            item.set_button_tooltip_text(0, 1, "Remove tag and descendants".into());
        }
    }
}

File: src/editor/controls/TagTree.cpp
//
// Created by iamoc on 20/03/2025.
//

#include "TagTree.h"

#include "../../TagDictionary.h"
#include <godot_cpp/classes/display_server.hpp>

using namespace octod::gameplay::tags;
using namespace octod::gameplay::tags::controls::TagTree;

static const char *TAG_PATH_META = "ggt_tag_dictionary_path";

void TagTree::_ready()
{
	Tree::_ready();

	connect("button_clicked", Callable::create(this, "_on_button_clicked"));
	connect("item_selected", Callable::create(this, "_on_item_selected"));
	connect("item_edited", Callable::create(this, "_on_tag_edited"));

	set_v_scroll_enabled(true);
	set_v_size_flags(SIZE_EXPAND_FILL);

	render_tree();
}

void TagTree::deselect_many_tags(const PackedStringArray &p_tags)
{
	for (int i = 0; i < p_tags.size(); i++)
	{
		if (const int64_t index = selected_tags.find(p_tags[i]); index >= 0)
		{
			selected_tags.remove_at(index);
		}
	}
}

void TagTree::deselect_tag(const String &p_tag)
{
	if (const int64_t index = selected_tags.find(p_tag); index >= 0)
	{
		selected_tags.remove_at(index);
	}
}

bool TagTree::get_editable() const
{
	return editable;
}

bool TagTree::get_selectable() const
{
	return selectable;
}

PackedStringArray TagTree::get_selected_tags() const
{
	return selected_tags;
}

TagDictionary *TagTree::get_tag_dictionary() const
{
	return tag_dictionary;
}

bool TagTree::is_path_selected(const String &p_tag_path) const
{
	for (int i = 0; i < selected_tags.size(); i++)
	{
		if (selected_tags[i].begins_with(p_tag_path))
		{
			return true;
		}
	}

	return false;
}

void TagTree::select_many_tags(const PackedStringArray &p_tags)
{
	for (int i = 0; i < p_tags.size(); i++)
	{
		if (const int64_t index = selected_tags.find(p_tags[i]); index < 0)
		{
			selected_tags.push_back(p_tags[i]);
		}
	}
}

void TagTree::select_tag(const String &p_tag)
{
	if (const int64_t index = selected_tags.find(p_tag); index < 0)
	{
		selected_tags.push_back(p_tag);
	}
}

void TagTree::set_editable(bool p_editable)
{
	editable = p_editable;
}

void TagTree::set_selectable(bool p_selectable)
{
	selectable = p_selectable;
}

void TagTree::set_tag_dictionary(TagDictionary *p_tag_dictionary)
{
	tag_dictionary = p_tag_dictionary;
}

void TagTree::_on_button_clicked(const Ref<TreeItem> &p_item, int p_column_id, int p_id, int p_mouse_button_index)
{
	const String &tag_path = get_meta(TAG_PATH_META, "");

	if (p_id == 1)
	{
		tag_dictionary->remove_path(tag_path);
		emit_signal("tag_path_removed", tag_path);
		render_tree();
	} else
	{
		const int child_count = get_child_count();
		const String &new_tag_path = "new_tag_" + itos(child_count);

		String new_tag_name;

		if (tag_path.is_empty())
		{
			new_tag_name = new_tag_path;
		} else
		{
			new_tag_name = tag_path + String(TagDictionary::SPLIT_CHAR) + new_tag_path;
		}

		if (tag_dictionary->add_tag(new_tag_name))
		{
			render_tree();
		}
	}
}

void TagTree::_on_item_selected()
{
	TreeItem *maybe_selected = get_selected();

	if (!maybe_selected || !tag_dictionary)
	{
		return;
	}

	const String &tag_path = maybe_selected->get_meta(TAG_PATH_META, "");
	DisplayServer::get_singleton()->clipboard_set(tag_path);

	if (selectable)
	{
		PackedStringArray added_tags;
		PackedStringArray found_tags = tag_dictionary->get_tags_from_path(tag_path);
		PackedStringArray removed_tags;

		bool is_checked = maybe_selected->is_checked(0);

		for (int64_t i = 0; i < found_tags.size(); i++)
		{
			if (is_checked)
			{
				if (const int64_t index = selected_tags.find(found_tags[i]); index >= 0)
				{
					selected_tags.remove_at(index);
					removed_tags.push_back(found_tags[i]);
				}
			} else
			{
				if (const int64_t index = selected_tags.find(found_tags[i]); index >= 0)
				{
					added_tags.push_back(found_tags[i]);
					selected_tags.push_back(found_tags[i]);
				}
			}
		}

		if (added_tags.size() > 0)
		{
			emit_signal("tag_added", added_tags);
		}

		if (removed_tags.size() > 0)
		{
			emit_signal("tag_removed", removed_tags);
		}
	}
}

void TagTree::_on_tag_edited()
{
}

void TagTree::render_dictionary()
{
}

void TagTree::render_tree()
{
}

void TagTree::set_tree_item_checkable(const Ref<TreeItem> &p_tree_item, const String &p_current_path, const String &p_key)
{
}

void TagTree::set_tree_item_editable_icon(Ref<TreeItem> &p_tree_item)
{
}

void TagTree::_bind_methods()
{
	/// methods binding
	ClassDB::bind_method(D_METHOD("deselect_many_tags", "tags"), &TagTree::deselect_many_tags);
	ClassDB::bind_method(D_METHOD("deselect_tag", "tag"), &TagTree::deselect_tag);
	ClassDB::bind_method(D_METHOD("get_editable"), &TagTree::get_editable);
	ClassDB::bind_method(D_METHOD("get_selectable"), &TagTree::get_selectable);
	ClassDB::bind_method(D_METHOD("get_selected_tags"), &TagTree::get_selected_tags);
	ClassDB::bind_method(D_METHOD("get_tag_dictionary"), &TagTree::get_tag_dictionary);
	ClassDB::bind_method(D_METHOD("is_path_selected", "tag_path"), &TagTree::is_path_selected);
	ClassDB::bind_method(D_METHOD("select_many_tags", "tags"), &TagTree::select_many_tags);
	ClassDB::bind_method(D_METHOD("select_tag", "tag"), &TagTree::select_tag);
	ClassDB::bind_method(D_METHOD("set_editable", "editable"), &TagTree::set_editable);
	ClassDB::bind_method(D_METHOD("set_selectable", "selectable"), &TagTree::set_selectable);
	ClassDB::bind_method(D_METHOD("set_tag_dictionary", "tag_dictionary"), &TagTree::set_tag_dictionary);
	ClassDB::bind_method(D_METHOD("_on_button_clicked", "item", "column_id", "id", "mouse_button_index"), &TagTree::_on_button_clicked);
	ClassDB::bind_method(D_METHOD("_on_item_selected"), &TagTree::_on_item_selected);
	ClassDB::bind_method(D_METHOD("_on_tag_edited"), &TagTree::_on_tag_edited);
	ClassDB::bind_method(D_METHOD("render_dictionary"), &TagTree::render_dictionary);
	ClassDB::bind_method(D_METHOD("render_tree"), &TagTree::render_tree);
	ClassDB::bind_method(D_METHOD("set_tree_item_checkable", "tree_item", "current_path", "key"), &TagTree::set_tree_item_checkable);
	ClassDB::bind_method(D_METHOD("set_tree_item_editable_icon", "tree_item"), &TagTree::set_tree_item_editable_icon);

	/// properties binding
	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "editable"), "set_editable", "get_editable");
	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "selectable"), "set_selectable", "get_selectable");
	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "tag_dictionary", PROPERTY_HINT_RESOURCE_TYPE, "24:17/TagDictionary"), "set_tag_dictionary", "get_tag_dictionary");

	/// signals binding
	ADD_SIGNAL(MethodInfo("tag_path_edited", PropertyInfo(Variant::STRING, "tag_path"), PropertyInfo(Variant::STRING, "new_tag_path")));
	ADD_SIGNAL(MethodInfo("tag_path_removed", PropertyInfo(Variant::STRING, "tag_path")));
	ADD_SIGNAL(MethodInfo("tags_added", PropertyInfo(Variant::ARRAY, "tags", PROPERTY_HINT_TYPE_STRING)));
	ADD_SIGNAL(MethodInfo("tags_removed", PropertyInfo(Variant::ARRAY, "tags", PROPERTY_HINT_TYPE_STRING)));
}
